node/src/utils/ds.rs

```diff
@@ -19,3 +29,134 @@ pub fn hash_set_fixed_size<T>(hashset: &HashSet<T>) -> usize {
+
+        let mut rng = sampling_rng(vec.len());
+        let sampled = vec
+            .as_slice()
+            .choose_multiple(&mut rng, SAMPLE_SIZE)
+            .map(|v| v.estimate_heap_size())
+            .sum();
+        base_size + scale_sample(vec.len(), sampled)
+    }
+}
+
+/// Extrapolate memory usage of a `HashMap` by from a random subset of `SAMPLE_SIZE` items.
+pub fn hashmap_sample<K, V>(map: &HashMap<K, V>) -> usize
+where
+    K: DataSize,
+    V: DataSize,
+{
+    if map.len() < SAMPLE_SIZE {
+        map.estimate_heap_size()
+    } else {
+        let base_size =
+            map.capacity() * (mem::size_of::<V>() + mem::size_of::<K>() + mem::size_of::<usize>());
+
+        let mut rng = sampling_rng(map.len());
+
+        let sampled = map
+            .iter()
+            .choose_multiple(&mut rng, SAMPLE_SIZE)
+            .into_iter()
+            .map(|(k, v)| k.estimate_heap_size() + v.estimate_heap_size())
+            .sum();
+
+        base_size + scale_sample(map.len(), sampled)
+    }
+}
+
+#[cfg(test)]
+#[allow(clippy::assertions_on_constants)] // used by sanity checks around `SAMPLE_SIZE`
+mod tests {
+    use std::collections::HashMap;
+
+    use datasize::DataSize;
+
+    use super::{hashmap_sample, vec_sample, SAMPLE_SIZE};
+
+    #[test]
+    fn vec_sample_below_sample_size() {
+        let data: Vec<Box<u32>> = vec![];
+
+        assert_eq!(vec_sample(&data), data.estimate_heap_size());
+
+        assert!(SAMPLE_SIZE > 3);
+        let data2: Vec<Box<u32>> = vec![Box::new(1), Box::new(2), Box::new(3)];
+
+        assert_eq!(vec_sample(&data2), data2.estimate_heap_size());
+    }
+
+    #[test]
+    fn vec_sample_above_sample_size() {
+        let num_items = SAMPLE_SIZE * 5;
+
+        // We make all items equal in size, so that we know the outcome of a random sampling.
+        let data: Vec<Vec<u32>> = (0..num_items)
+            .map(|_| vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
+            .collect();
+
+        assert_eq!(vec_sample(&data), data.estimate_heap_size());
+    }
+
+    #[test]
+    fn hashmap_sample_below_sample_size() {
+        let data: HashMap<u32, Box<u32>> = HashMap::new();
+
+        assert_eq!(hashmap_sample(&data), data.estimate_heap_size());
+
+        assert!(SAMPLE_SIZE > 3);
+        let mut data2: HashMap<u32, Box<u32>> = HashMap::new();
+        data2.insert(1, Box::new(1));
+        data2.insert(2, Box::new(2));
+        data2.insert(3, Box::new(3));
+
+        assert_eq!(hashmap_sample(&data2), data2.estimate_heap_size());
+    }
+
+    #[test]
+    fn hashmap_sample_above_sample_size() {
+        let num_items = SAMPLE_SIZE * 5;
+
+        // We make all items equal in size, so that we know the outcome of a random sampling.
+        let data: HashMap<usize, Vec<u32>> = (0..num_items)
+            .map(|idx| (idx, vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 10]))
+            .collect();
+
+        assert_eq!(hashmap_sample(&data), data.estimate_heap_size());
+    }
```

Comment:
Isn't the point of this "sampling" functions to provide a _good enough_ estimate where it's expensive to calculate the exact value? That's not the case for primitive types though, right?

Reply:
The exact calculation would iterate over all entries in the hash map, and sum up their sizes, whereas the estimate only takes the average of a sample, and multiplies it by the number of entries.

Even for a hash map of vectors, that should be faster than the full iteration.

Reply:
I get that. My point is that this particular test is not very useful in a sense that it doesn't test what's been added here. It doesn't verify how many iterations we did so `hashmap_sample` could be doing exactly what you described.

Reply:
Yes, I think that test is just to verify that the new method is correct, not that it's faster. But I'll let @marc-casperlabs comment on that.

Reply:
And it's trivial for the primitive types since its size is constant.

Reply:
But in this case it's not trivial because it's a map of _vectors_ of `u32`s, not a map of `u32`s.